### ISSUE TITLE
Crash while updating Adguard

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/AdguardForAndroid#1086]
```
Thread: Thread[main,5,main]; Uncaught exception:

java.lang.RuntimeException: Error receiving broadcast Intent { act=android.intent.action.DOWNLOAD_COMPLETE flg=0x10 pkg=com.adguard.android (has extras) } in com.adguard.android.service.am@40d6306
	at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:891) ~[na:0.0]
	at android.os.Handler.handleCallback(Handler.java:739) ~[na:0.0]
	at android.os.Handler.dispatchMessage(Handler.java:95) ~[na:0.0]
	at android.os.Looper.loop(Looper.java:148) ~[na:0.0]
	at android.app.ActivityThread.main(ActivityThread.java:5437) ~[na:0.0]
	at java.lang.reflect.Method.invoke(Native Method) ~[na:0.0]
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) ~[na:0.0]
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) ~[na:0.0]
	at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:102) ~[na:na]
Caused by: java.lang.SecurityException: Permission Denial: reading com.android.providers.downloads.DownloadProvider uri content://downloads/all_downloads/95 from pid=28288, uid=10112 requires android.permission.ACCESS_ALL_DOWNLOADS, or grantUriPermission()
	at android.os.Parcel.readException(Parcel.java:1620) ~[na:0.0]
	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183) ~[na:0.0]
	at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel(DatabaseUtils.java:146) ~[na:0.0]
	at android.content.ContentProviderProxy.openTypedAssetFile(ContentProviderNative.java:692) ~[na:0.0]
	at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1107) ~[na:0.0]
	at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:944) ~[na:0.0]
	at android.content.ContentResolver.openFileDescriptor(ContentResolver.java:797) ~[na:0.0]
	at android.content.ContentResolver.openFileDescriptor(ContentResolver.java:751) ~[na:0.0]
	at android.app.DownloadManager.openDownloadedFile(DownloadManager.java:1010) ~[na:0.0]
	at com.adguard.android.service.am.onReceive(Unknown Source) ~[na:0.0]
	at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:881) ~[na:0.0]
	... 8 common frames omitted
```

**Comment by ameshkov** [comment]
At least we should catch it

**Comment by Revertron** [comment]
Changing permission string from `
<uses-permission android:name="android.permission.ACCESS_ALL_DOWNLOADS"/>` to `<uses-permission android:maxSdkVersion="21" android:name="android.permission.ACCESS_ALL_DOWNLOADS"/>` Eliminates the string `Not granting permission android.permission.ACCESS_ALL_DOWNLOADS to package com.adguard.android (protectionLevel=2 flags=0x3818be46)` from the logs while updating/installing Adguard.
Simply put, we are telling Android that this permission is used like on the old version of OS :)

**Comment by TPS** [comment]
I don't know if this is same crash, but I just filed bug report (UserID should still be 1122984, I think) for today's update via Amazon Appstore:
![screenshot_20170309-133956 01](https://cloud.githubusercontent.com/assets/185902/23765815/7451941a-04d0-11e7-9e4e-99bbfce10321.png)

**Comment by ameshkov** [comment]
More likely to be an issue with downloaded file, it might be corrupted. Does it help to tap update second time?

**Comment by TPS** [comment]
The app downloaded & installed fine, but, when running it gives this & crashes.  Are you asking me to uninstall & redownload?

**Comment by ameshkov** [comment]
Wow, I then have no idea what's going on, really:)

**Comment by ameshkov** [comment]
I wonder could it be damaged somehow with the amazon DRM?

**Comment by TPS** [comment]
Does my bug report help @ all?

**Comment by ameshkov** [comment]
I am pretty sure it should be filed as a separate bug report. But only if it is reproducible:)

**Comment by TPS** [comment]
Well, I've had it happen 2× in 10 minutes, but I've no idea how I did it.  I'll play w/ it & see.

**Comment by Bohdan-SUP** [comment]
The log, just in case - 
[debug (2).txt](https://github.com/AdguardTeam/AdguardForAndroid/files/832058/debug.2.txt)

**Comment by ameshkov** [comment]
Found this:
https://www.amazon.com/gp/help/customer/forums/kindleqna/ref=cs_hc_k_tv?ie=UTF8&forumID=Fx1FI6JDSFEQQ7V&cdThread=TxMJCTY32PBJPI

**Comment by TPS** [comment]
@Bohdan-SUP Thanks!

@ameshkov Well, this isn't a Kindle nor any other kind of Amazon device, & my wifi's been solid, so it doesn't seem to apply. 🤷

**Comment by ameshkov** [comment]
Done

**Comment by ameshkov** [comment]
Guys, let's add ACCESS_ALL_DOWNLOADS after all, there are more reports on this issue

**Comment by TPS** [comment]
Further update from my side (sorry for not reporting earlier):  Had same problem w/ last ß update, too.  Killing app & clearing app cache _might_ have fixed it, but it definitely worked after reboot. ☺

**Comment by ameshkov** [comment]
@TPS something is really not okay with Amazon DRM:)

**Comment by Revertron** [comment]
Fixed.

**Comment by nkartyshov** [comment]
I tried update AG with 2.9.111 to current version and app crash with exception:
> android.os.FileUriExposedException: file:///storage/emulated/0/Android/data/com.adguard.android/cache/adguard.apk exposed beyond app through Intent.getData()

**Comment by ameshkov** [comment]
@nkartyshov exception text is a bit different from this issue, let's file a new bug report instead.
